modules/vision/src/pose-estimation/vpPoseRGBD.cpp

```diff
@@ -248,86 +252,231 @@ bool validPose(const vpHomogeneousMatrix& cMo) {
+                      point3d.size(), corners.size()));
+  }
+  std::vector<vpPoint> pose_points;
+  if (confidence_index != NULL) {
+    *confidence_index = 0.0;
+  }
 
-      // Plane equation
-      vpColVector plane_equation, centroid;
-      double normalized_weights = 0;
-      estimatePlaneEquationSVD(points_3d, plane_equation, centroid, normalized_weights);
+  for (size_t i = 0; i < point3d.size(); i++) {
+    std::vector<vpPoint> tagPoint3d = point3d[i];
+    for (size_t j = 0; j < tagPoint3d.size(); j++) {
+      pose_points.push_back(tagPoint3d[j]);
+    }
+  }
 
-      for (size_t j = 0; j < corners.size(); j++) {
-          const vpImagePoint& imPt = corners[j];
+  //total Area of the polygon to estimate confidence
+  double totalArea = 0.0;
+  std::vector<vpPolygon> vec_polygon(corners.size());
+  std::vector<double> tag_points_3d;
+
+  for (size_t i = 0; i < vec_polygon.size(); i++) {
+    vpPolygon polygon(corners[i]);
+    vpRect bb = polygon.getBoundingBox();
+    totalArea += polygon.getArea();
+    std::vector<double> points_3d;
+
+    unsigned int top = static_cast<unsigned int>(std::max(0, static_cast<int>(bb.getTop())));
+    unsigned int bottom = static_cast<unsigned int>(
+        std::min(static_cast<int>(depthMap.getHeight()) - 1, static_cast<int>(bb.getBottom())));
+    unsigned int left = static_cast<unsigned int>(std::max(0, static_cast<int>(bb.getLeft())));
+    unsigned int right =
+        static_cast<unsigned int>(std::min(static_cast<int>(depthMap.getWidth()) - 1, static_cast<int>(bb.getRight())));
+
+    points_3d.reserve((bottom - top) * (right - left));
+    for (unsigned int idx_i = top; idx_i < bottom; idx_i++) {
+      for (unsigned int idx_j = left; idx_j < right; idx_j++) {
+        vpImagePoint imPt(idx_i, idx_j);
+        if (depthMap[idx_i][idx_j] > 0 && polygon.isInside(imPt)) {
           double x = 0, y = 0;
           vpPixelMeterConversion::convertPoint(colorIntrinsics, imPt.get_u(), imPt.get_v(), x, y);
-          double Z = computeZMethod1(plane_equation, x, y);
-          if (Z < 0) {
-              Z = -Z;
-          }
-          p.push_back(vpPoint(x*Z, y*Z, Z));
-
-          pose_points[j].set_x(x);
-          pose_points[j].set_y(y);
+          double Z = depthMap[idx_i][idx_j];
+          points_3d.push_back(x * Z);
+          points_3d.push_back(y * Z);
+          points_3d.push_back(Z);
+        }
       }
+    }
+    tag_points_3d.insert(tag_points_3d.end(), points_3d.begin(), points_3d.end());
+  }
 
-      for (size_t i = 0; i < point3d.size(); i ++) {
-        q.push_back(point3d[i]);
+  unsigned int nb_points_3d = static_cast<unsigned int>(tag_points_3d.size() / 3);
+
+  if (nb_points_3d > 4) {
+    std::vector<vpPoint> p, q;
+
+    // Plane equation
+    vpColVector plane_equation, centroid;
+    double normalized_weights = 0;
+    estimatePlaneEquationSVD(tag_points_3d, plane_equation, centroid, normalized_weights);
+
```

Comment:
To adress the more generic case where all the tags don't belong to the same plane, an idea would be to compute one plane for each polygon. Thus `estimatePlaneEquationSVD()` should be called for each polygon.

Can you try to modify the code...

Reply:
Thanks, I'll modify it and test two solutions carefully.

Reply:
Ah, actually in my case, we also put the tags in same plane for another object. So should I write another function or add an argument to distinguish the case?

Reply:
In a first attempt, you can add a parameter like `bool coplanar_points` that could be useful to compare results. I hope that for all your use cases the generic case (`coplanar_points = false`) would give good results.